### ISSUE TITLE
Building `cuda.bindings` in parallel might fail non-deterministically

### ISSUE THREAD

**Issue by leofang** [NVIDIA/cuda-python#271]
Our new CI (#267) immediately exposes a longstanding, widely known `setuptools`/`distutils` issue when `PARALLEL_LEVEL` is set, xref:
- https://github.com/NVIDIA/cuda-python/actions/runs/12192916020/job/34014323275#step:4:2187
```
    D:\a\cuda-python\cuda-python\cuda_bindings\cuda\bindings\_bindings\loader.cpp : fatal error C1083: Cannot open compiler generated file: 'D:\a\cuda-python\cuda-python\cuda_bindings\build\temp.win-amd64-cpython-310\Release\cuda\bindings\_bindings\loader.obj': Permission denied
```
- https://github.com/pypa/setuptools/issues/3119
- https://github.com/pypa/setuptools/issues/3942
- https://github.com/pypa/setuptools/issues/2442
- scipy/scipy#7112

**Comment by leofang** [comment]
tl;dr for the above threads is: The way that we build/link the same .c/.cpp file to multiple shared extension modules, i.e.,
https://github.com/NVIDIA/cuda-python/blob/60f93627b3c79589879204c84ed3556a4f3ea12f/cuda_bindings/setup.py#L267-L268
is the root cause, as it triggers setuptools's race condition (which is a bug that would unlikely be fixed).

**Comment by leofang** [comment]
Bumping this to P0 as it now shows up in virtually every single (build) CI runs. @vzhurba01 please prioritize for fixing issues caught by the CI. Thanks!

**Comment by leofang** [comment]
> Bumping this to P0 as it now shows up in virtually every single (build) CI runs.

I notice the situation gets worse after CI 2.0 was merged. Now when we rerun the failed jobs, not only the failed Windows build jobs are rerun, but also all downstream (test/doc) jobs, because as noted earlier there's no way in GHA to declare dependency on specific matrix elements of a job, only on the whole job.
![Image](https://github.com/user-attachments/assets/f47c1a79-08c3-4394-83b4-46a6d89941dd)

**Comment by leofang** [comment]
Looks like we fixed it! The build workflow has been run many times today, but I haven't observed any flaky failures happening.
